### PR TITLE
Auto-deploy to production after test deployment

### PR DESCRIPTION
**Pull request by bickelj**
Without this change, someone needs to verify that the automated deployment to the test environment succeeded and then manually type two or three commands to deploy to production. At the moment, there is only one person who has ever done such production deployments.

With this change, however, the deployment to the test environment gets verified by visiting a URL that exposes the current deployed version. When the body of a request sent to that URL returns the same string as the version (tag) sent to this deployment job, this action can safely conclude that the test deployment has succeeded. How and why can it make such a haughty conclusion? Several reasons:

* The `deploy.sh` script saves the new version only on success.
* The server in the `reverse-proxy` container serves the version that the `deploy.sh` script wrote.
* The `reverse-proxy` container does not start until the `database` and `web` containers are running according to docker health checks.

In other words, accessing this version string implies that the state of the environment is OK. If the version string matches in the test environment, this gives some confidence that auto-deploying to the production environment is relatively safe. Therefore it auto-deploys to production. This means that merging pull requests in the `service` repository is sufficient to cause automatic deployments to production and at the same time the production deployment is safe due to the same steps having run for a deployment to test. It means no more asking for some person to deploy to production anymore!

Since "send tag to machine" is indecipherable, the action is renamed.

Issue #106 Auto-deploy to production

### CODE REVIEW

**Comment by bickelj** [comment]
Nice, https://github.com/PhilanthropyDataCommons/deploy/actions/runs/8426881659

File: .github/workflows/trigger-deployments.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
+name: Trigger deployment on test then prod
+
+on:
+  push:
```
Comment (slifty) [inline .github/workflows/trigger-deployments.yml:4]: As written, this will trigger on every push to ANY branch, right?

I know we don't really do much dev work in here, but we probably should update this to only happen on pushes to main to give us more control.

File: .github/workflows/trigger-deployments.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    - run: scp -v -i ~/ssh_id -o UserKnownHostsFile=~/my_known_hosts -p ~/tag_to_deploy build@${{ secrets.MACHINE_ADDRESS }}:/home/build/deployment/tag_to_deploy
+    - uses: artiz/poll-endpoint@1.0.2
+      with:
+        url: ${{ secrets.DEPLOYED_VERSION_URL }}
+        method: GET
+        expectStatus: 200
+        expectBody: "${{ github.ref_name }}"
+        timeout: 600000
+        interval: 20000
+  trigger-production-deployment:
```
Comment (slifty) [inline .github/workflows/trigger-deployments.yml:43]: This looks identical to `trigger-test-deployment` -- do they both push to the same place?  What's the point of having both?  Is there a way to DRY this at all?

**Review by slifty — DISMISSED** [review]
I *think* my reading is correct about the trigger condition -- let's make the trigger be "push to main" rather than a general "push".

https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushbranchestagsbranches-ignoretags-ignore

File: .github/workflows/trigger-deployments.yml (hunk 1 above)
Comment (bickelj) [reply .github/workflows/trigger-deployments.yml:4]: The tags filter below it should mean it effectively only runs when a tag is pushed. Maybe there is a better way, though:
```
on:
  push:
    tags:
    - '*'
```

A quick search reveals [a SO post suggesting the same](https://stackoverflow.com/questions/61891328/trigger-github-action-only-on-new-tags).

This part of the code had been working fine [in the previous incarnation](https://github.com/PhilanthropyDataCommons/deploy/pull/107/files#diff-43b13329f8257d4e946cab27f88b61d3b8738929640b86496e7169dc8da9f871L5).

Comment (bickelj) [reply .github/workflows/trigger-deployments.yml:4]: In other words, in this repo, we don't want to auto-push anything that gets merged to main, only when a tag gets pushed. On a push to main in the `service` repo, this will cause a tag to get pushed to this repo. Maybe that design is up for review too but it's been working well enough for the past couple years.

File: .github/workflows/trigger-deployments.yml (hunk 2 above)
Comment (bickelj) [reply .github/workflows/trigger-deployments.yml:43]: @slifty You are right that the body is identical and it would be better to make it DRY. They are identical except for the environment. The environment substitutes the appropriate variables. I find [evidence that we can break it out](https://docs.github.com/en/actions/using-workflows/reusing-workflows). I think the only downside is this will mean another `.yml` file separate from this one. But it is probably worth it because yes, it's exactly the same code with different environment variable values.

Comment (bickelj) [reply .github/workflows/trigger-deployments.yml:43]: @slifty And no, they do not push to the same place, one pushes to the test environment (think "deploy preview") and the other pushes to the production environment.

Comment (bickelj) [reply .github/workflows/trigger-deployments.yml:43]: @slifty The point of having both is to deploy to the test environment prior to deploying to the production environment. If everything works during test deployment such that the services are up and responsive, this increases confidence and reduces the risk of something going wrong when deploying to production.

**Comment by bickelj** [comment]
Hmm, trying to wrap my head [around (a solution to) these limitations](https://docs.github.com/en/actions/using-workflows/reusing-workflows#limitations):
> Any environment variables set in an env context defined at the workflow level in the caller workflow are not propagated to the called workflow. For more information, see "Variables" and "Contexts."
>
> Similarly, environment variables set in the env context, defined in the called workflow, are not accessible in the env context of the caller workflow. Instead, you must use outputs of the reusable workflow. For more information, see "Using outputs from a reusable workflow."

**Comment by bickelj** [comment]
Likewise, [these limitations](https://docs.github.com/en/actions/learn-github-actions/contexts#secrets-context)
> The secrets context is not available for composite actions due to security reasons. If you want to pass a secret to a composite action, you need to do it explicitly as an input.

Maybe making this a bash script would do the trick.

File: .github/workflows/trigger-deployments.yml (hunk 1 above)
Comment (slifty) [reply .github/workflows/trigger-deployments.yml:4]: @bickelj could we have BOTH -- e.g. "the push has to have a tag AND be to main"?

I think what separates this from the old way is now we're talking about pushing to production so stricter scrutiny is useful.  I *think* that would still work with the existing model, right?

File: .github/workflows/trigger-deployments.yml (hunk 2 above)
Comment (slifty) [reply .github/workflows/trigger-deployments.yml:43]: Ahh got it I didn't realize the environment dictated that -- makes sense!

DRY would be nice to have, but I defer to you on what makes sense.  Thanks for the explanation!

File: .github/workflows/trigger-deployments.yml (hunk 1 above)
Comment (bickelj) [reply .github/workflows/trigger-deployments.yml:4]: @slifty By restricting it to a tag push already, we decouple deployments from code changes (edit: in this `deploy` repository, I mean). To push a tag is a special action. The tag references a commit much like a branch does (internally I'm not sure there's much of a difference). If the tag name does not conform to the scripts' expected syntax, then a deployment ~~won't happen~~ shouldn't succeed (see [validation in deploy.sh](https://github.com/PhilanthropyDataCommons/deploy/blob/d066112b1e7c1a32bbfbe3e53a59d02652ca448b/deploy.sh#L92)). Perhaps we need to protect against accidental tag pushes. I guess I can conceive of a tag referencing some branch somewhere instead of the main branch. But the main branch can change too, away from a previously-tagged commit. Are you saying we need a rule somewhere that "only on a tag that references a presently-in-the-main-branch commit this flow should run"? That makes sense but I'm not sure how to do that yet. I can look into it.

Comment (bickelj) [reply .github/workflows/trigger-deployments.yml:4]: Right now, the auto-tag has happened only on the main branch, see https://github.com/PhilanthropyDataCommons/deploy/blob/main/.github/workflows/update-service-image.yml - but to your point, perhaps that needs to be strictly limited to main as well.

Edit: I'm not sure how that can run against anything but main, actually. Hmm.

Edit2: see [comment on the main thread](https://github.com/PhilanthropyDataCommons/deploy/pull/107#issuecomment-2023363866)

File: .github/workflows/trigger-deployments.yml (hunk 2 above)
Comment (bickelj) [reply .github/workflows/trigger-deployments.yml:43]: OK, I made it DRYer than before by putting the script steps into an actual bash script. This still causes some indirection but it should be fairly easy to read and follow.

Some benefits include
* the script is now in the repo,
* documented, and
* `shellcheck`ed when changed (#80 #103).

The downside is the one step of indirection, the uncertainty that this will actually work when pushed (but we can fix it), and there are still a couple of steps repeated (check out and poll the endpoint).

What do you think?

**Comment by bickelj** [comment]
It would be good to strictly deploy what is in the main branch. Right now that's what has happened, but conceivably a tag could be added to any old commit and it could get deployed to prod: not good! This might help: https://github.com/rickstaa/action-contains-tag

File: .github/workflows/trigger-deployments.yml (hunk 1 above)
Comment (bickelj) [reply .github/workflows/trigger-deployments.yml:4]: I used [action-contains-tag](https://github.com/rickstaa/action-contains-tag) as an additional protection for production deployments. So now production deployments require both the test deployment to succeed *and* the pushed tag to be verified as part of the main branch. I could imagine a scenario where perhaps we'd like to push a tag on a branch that we'd like deployed to the test environment but not production which is why I didn't add the same check on the deployment to test.

**Comment by bickelj** [comment]
FWIW, dry runs locally using https://github.com/nektos/act have helped a lot during these edits! Thank you, @nektos!

**Comment by bickelj** [comment]
@slifty Regarding pushes to main, I would prefer to be able to merge script edits here in this deploy repo without triggering a deployment of the software to production, hence keeping deploys here based on pushing a tag. But I have added a check that a pushed tag is part of the main branch prior to triggering a production deployment because we definitely do not want non-main code going to prod!

**Review by slifty — APPROVED** [review]
Looks good!  I didn't meant to complicate with the main branch check, and wanna just say if you DONT think that's useful please do take it out.

**Comment by bickelj** [comment]
> Looks good! I didn't meant to complicate with the main branch check, and wanna just say if you DONT think that's useful please do take it out.

I think it's very useful. We would not want to deploy any build to production that wasn't built from the main branch, for sure!
